x/stake/pool_test.go

```diff
@@ -1 +1,354 @@
+			}
+			msg := fmt.Sprintf("candidate with pubkey %s, %d assets, %d liabilities, and %d delegatorShareExRate",
+				cand.PubKey, cand.Assets.Evaluate(), cand.Liabilities.Evaluate(), cand.delegatorShareExRate().Evaluate())
+			p, cand, tokens := p.candidateRemoveShares(cand, shares)
+			c[index] = cand
+			t += tokens
+			msg = fmt.Sprintf("Removed %d shares from %s", shares.Evaluate(), msg)
+			return p, c, t, msg
+		},
+	}
+	r.Shuffle(len(operations), func(i, j int) {
+		operations[i], operations[j] = operations[j], operations[i]
+	})
+	return operations[0]
+}
+
+// ensure invariants that should always be true are true
+func assertInvariants(t *testing.T, pA Pool, cA Candidates, tA int64, pB Pool, cB Candidates, tB int64, msg string) {
+
+	// total tokens conserved
+	require.Equal(t, pA.UnbondedPool+pA.BondedPool+tA, pB.UnbondedPool+pB.BondedPool+tB)
+
+	// nonnegative shares
+	require.Equal(t, pB.BondedShares.LT(sdk.ZeroRat), false)
+	require.Equal(t, pB.UnbondedShares.LT(sdk.ZeroRat), false)
+
+	// nonnegative ex rates
+	require.Equal(t, pB.bondedShareExRate().LT(sdk.ZeroRat), false,
+		"Applying operation \"%s\" resulted in negative bondedShareExRate: %d",
+		msg, pB.bondedShareExRate().Evaluate())
+	require.Equal(t, pB.unbondedShareExRate().LT(sdk.ZeroRat), false,
+		"Applying operation \"%s\" resulted in negative unbondedShareExRate: %d",
+		msg, pB.unbondedShareExRate().Evaluate())
+	bondedSharesHeld := sdk.ZeroRat
+	unbondedSharesHeld := sdk.ZeroRat
+
+	for _, candidate := range cA {
+
+		// nonnegative ex rate
+		require.Equal(t, false, candidate.delegatorShareExRate().LT(sdk.ZeroRat),
+			"Applying operation \"%s\" resulted in negative candidate.delegatorShareExRate(): %s (candidate.PubKey: %s)",
+			msg, candidate.delegatorShareExRate(), candidate.PubKey)
+
+		// nonnegative assets / liabilities
+		require.Equal(t, false, candidate.Assets.LT(sdk.ZeroRat),
+			"Applying operation \"%s\" resulted in negative candidate.Assets: %d (candidate.Liabilities: %d, candidate.PubKey: %s)",
+			msg, candidate.Assets.Evaluate(), candidate.Liabilities.Evaluate(), candidate.PubKey)
+
+		require.Equal(t, false, candidate.Liabilities.LT(sdk.ZeroRat),
+			"Applying operation \"%s\" resulted in negative candidate.Liabilities: %d (candidate.Assets: %d, candidate.PubKey: %s)",
+			msg, candidate.Liabilities.Evaluate(), candidate.Assets.Evaluate(), candidate.PubKey)
+
+		if candidate.Status == Bonded {
+			bondedSharesHeld = bondedSharesHeld.Add(candidate.Assets)
+		} else {
+			unbondedSharesHeld = unbondedSharesHeld.Add(candidate.Assets)
+		}
+	}
+
+	// shares outstanding = total shares held by candidates, both bonded and unbonded
+	require.Equal(t, bondedSharesHeld, pB.BondedShares)
+	require.Equal(t, unbondedSharesHeld, pB.UnbondedShares)
+}
+
+// run random operations in a random order on a random state, assert invariants hold
+func TestIntegrationInvariants(t *testing.T) {
+	r := rand.New(rand.NewSource(int64(42)))
```

Comment:
This is generating one deterministic number each time and reusing it everywhere... what you want is two instances of rand within each loop: `rand.New(rand.NewSource(time.Now().UnixNano()))`

.... making this fix now